### ISSUE TITLE
Opt-in or info-level notification about type narrowing potentially being wrong

### ISSUE THREAD

**Issue by lengau** [microsoft/pyright#10492]
**Is your feature request related to a problem? Please describe.**

Consider this code:

```python
def set_none(container: dict[str, str | None]) -> None:
    container["key"] = None


def tester(x: dict[str, str | None]):
    assert x["key"] is not None
    set_none(x)
    "".join([x["key"]])  # TypeError: sequence item 0: expected str instance, NoneType found


parent: dict[str, str | None] = {"key": "value"}
tester(parent)

```

Pyright treats this as valid even though it will always result in a TypeError. While changing the behaviour so it treats it as an error is probably unwise (see https://github.com/astral-sh/ty/issues/164#issuecomment-2902101002), it would be nice if the LSP would note this.

**Describe the solution you’d like**

If a narrowed type is assumed after a function call or similar could reset that type narrowing, it would be nice if we could be made aware of that. (Ideally it would notify only if the called code actually does mutate the relevant attribute/item, but that may not be feasible.)

**Comment by erictraut** [comment]
Invalidating narrowing after every function call would result in a really bad type checking experience. Any call can theoretically invalidate any type narrowing assumptions, but in practice this rarely happens in well-structured code.

For comparison, the TypeScript compiler is the same as pyright in this regard. As Carl pointed out in the thread in the ty issue tracker, pyre tried adding something like this, and it was very unpopular. Let's learn from that failed experiment.

If we were to add a feature like this, it would need to be off by default even in strict mode. In general, I don't like adding checks to pyright that cannot be enabled by default in strict mode. They tend to be hard to discover and therefore get very little use.

For these reasons, I think it's unlikely that we will add something like this to pyright. I might be persuaded to revisit this decision if there was _significant_ feedback from pyright users that it was broadly desired, but the bar is very high here given the history and precedent that we can draw upon.
